### PR TITLE
Merge `quarkus3` branch into main

### LINKED ISSUE #16246: Upgrade to Quarkus 3
### Description

For Keycloak 22, we would like to upgrade Quarkus to version 3.x, which probably will require more tasks to get done. We'd like to use the migration script provided by the Quarkus team for upgrading included dependencies in projects and settings proper versions for them. 

### Acceptance criteria
- Migrate from JavaEE to JakartaEE and its dependencies (except `adapters` modules). 
- Upgrade to Quarkus 3.x
- Upgrade to Hibernate ORM 6.x
- Passing the testsuite

Hibernate ORM 6 upgrade: #16127      

###  Related Issues and PRs
- #19161
- #17664
- #19571
- https://github.com/keycloak/keycloak/pull/19760

### Related Issues
- #16247
- #16251
- #19084
- #19069
- #19162
- #19573
- https://github.com/keycloak/keycloak/issues/20384

### Follow-up tasks
- #19299
- https://github.com/keycloak/keycloak/issues/19834
- https://github.com/keycloak/keycloak/issues/19809

### PR DESCRIPTION
**Pull request by mabartos**
@keycloak/cloud-native @keycloak/store @keycloak/maintainers 
Could you please check this PR?

Individual commits were progressively reviewed.

For more information, check issue: #16246.

Known follow-up tasks (all of them tracked in #16246):

- https://github.com/keycloak/keycloak/issues/16251
- https://github.com/keycloak/keycloak/issues/19299
- https://github.com/keycloak/keycloak/issues/19573
- https://github.com/keycloak/keycloak/issues/19834
- https://github.com/keycloak/keycloak/issues/19809

### CODE REVIEW

**Review by ahus1 — APPROVED** [review]
I've been working regularly with this branch and from my store team's perspective it is ok to be merged. Thanks to everyone involved in this. While there is some more work to do, it is great to see the tests green.

File: saml-core-api/pom.xml
Comment (vmuzikar) [inline saml-core-api/pom.xml:62]: @pskopek Is this fine for prod?

**Review by vmuzikar — APPROVED** [review]
I did a sanity check of each file and LGTM. Wonderful work from all involved!

**Review by mposolda — APPROVED** [review]
Great work to everyone involved in this PR!

@vmuzikar I am approving as overally it looks good. I did not reviewed too deeply, but as you mentioned, I rely on the fact that parts of this PR were already reviewed before. Let me know if you prefer me to look at some concrete parts in more details.

**Comment by abstractj** [comment]
Hi @mabartos, I could not comment on the code because my browser is crashing to open it on GH. But I would suggest updating to 3.0.1 (https://github.com/quarkusio/quarkus/releases/tag/3.0.1.Final). The Quarkus team released it a few hours ago, I can run our security scanner to identify CVEs after the change.

**Comment by mabartos** [comment]
@abstractj Thanks for the message. Yes, we could update it as I think also the Hibernate ORM should be updated.

**Review by pskopek — APPROVED** [review]
Changes to POM files looks fine. After this PR is merged we need to review some of adapers modules generated for Jakarta transition (I will do it).

**Review by pedroigor — APPROVED** [review]
Great work!

Anything missing/fix/improvement we can do as a follow-up.

**Review by abstractj — APPROVED** [review]
@mabartos approving it regardless, we can switch to the next release of Quarkus after the merge. Great job, and thank you.

**Review by hmlnarik — APPROVED** [review]
Approving per prior approvals during the development of `quarkus3` branch.

**Review by ssilvert — APPROVED** [review]

File: integration/admin-client/src/main/java/org/keycloak/admin/client/ClientBuilderWrapper.java
Comment (stianst) [inline integration/admin-client/src/main/java/org/keycloak/admin/client/ClientBuilderWrapper.java:1]: As `admin-client` is now updated to use Jakarta that leaves us without an admin client for JavaEE it seems? Is that correct? I could imagine that would be disruptive to those still using the admin client on JavaEE, and also what is then the purpose of `admin-client-jakarta`?

**Review by stianst — COMMENTED** [review]
See comment with regards to `admin-client`. Do we no longer have an admin-client for JavaEE with this PR?

Also, with changes to `saml-core` and `saml-core-api` does that mean we no longer have support for SAML adapters on JavaEE? That's probably okay, but just wanted to double check on this one.

What are the plans with regard to quickstarts? I presume quite a few will need updating/fixing after this is merged?

What are the plans around documentation? We would at the very least need release notes and migration guide for this change, but we may also need to review documentation throughout for any code examples or anything else that is referring to JavaEE when it should refer to Jakarta.

**Comment by mabartos** [comment]
With @miquelsi, we've encountered an issue in the build pipeline where WildFly adapters are not explicitly built. Until now, WildFly adapters were fetched from the remote M2 repository and not from the local one; it fetches nightly WildFly adapters. It's fine because there are no changes in the WildFly adapters in comparison to the main/quarkus3. 

This can be done as a follow-up task. @miquelsi WDYT?
https://github.com/keycloak/keycloak/pull/19935

File: integration/admin-client/src/main/java/org/keycloak/admin/client/ClientBuilderWrapper.java
Comment (mabartos) [reply integration/admin-client/src/main/java/org/keycloak/admin/client/ClientBuilderWrapper.java:1]: @stianst Thanks for your comment. We have `admin-client`, which should support JavaEE, and `admin-client-jakarta', which should support JakartaEE. I think the conversion from JavaEE -> JakartaEE was done unintentionally and could be reverted back. 

Could be done in a follow-up task?

**Comment by mabartos** [comment]
> What are the plans around documentation? We would at the very least need release notes and migration guide for this change, but we may also need to review documentation throughout for any code examples or anything else that is referring to JavaEE when it should refer to Jakarta.

Yes, we definitely need release notes and a migration guide. This follow-up task should cover that: https://github.com/keycloak/keycloak/issues/16251. I've updated the issue in order to cover all docs tasks.

> What are the plans with regard to quickstarts? I presume quite a few will need updating/fixing after this is merged?

I've created a task for integration QSs with Quarkus 3. See https://github.com/keycloak/keycloak/issues/19980

**Comment by mabartos** [comment]
> Also, with changes to saml-core and saml-core-api does that mean we no longer have support for SAML adapters on JavaEE? That's probably okay, but just wanted to double check on this one.

We're using `saml-core` and `saml-core-api` for adapters (JavaEE), but also for the server-side modules (JakartaEE). These SAML modules contain both JavaEE and JakartaEE dependencies and it should be possible to use them for both cases. AFAIK, there are no points in the code where it's necessary to require JavaEE or JakartaEE explicitly; especially for adapters. Thus, I presume we can leave it as is. 

We covered that behavior for the WildFly adapter in the testsuite, which supports JavaEE, and everything works as expected.

**Comment by vmuzikar** [comment]
Thanks for all the reviews and the valuable feedback!

+1 for addressing the concerns as follow-ups if possible as suggested by @mabartos. Given the size of the PR and high conflicts potential, and that other work is blocked by it (e.g. SmallRye keystore), I'd like to ask you to consider prioritizing merging it.

File: integration/admin-client/src/main/java/org/keycloak/admin/client/ClientBuilderWrapper.java
Comment (stianst) [reply integration/admin-client/src/main/java/org/keycloak/admin/client/ClientBuilderWrapper.java:1]: Follow-up works fine for me. I've added milestone 22 to that issue + the priority/critical label.

**Comment by stianst** [comment]
> > What are the plans around documentation? We would at the very least need release notes and migration guide for this change, but we may also need to review documentation throughout for any code examples or anything else that is referring to JavaEE when it should refer to Jakarta.
> 
> Yes, we definitely need release notes and a migration guide. This follow-up task should cover that: #16251. I've updated the issue in order to cover all docs tasks.
> 
> > What are the plans with regard to quickstarts? I presume quite a few will need updating/fixing after this is merged?
> 
> I've created a task for integration QSs with Quarkus 3. See #19980

Thanks, I've added milestone 22 and priority/critical to both those issues

**Comment by stianst** [comment]
> > Also, with changes to saml-core and saml-core-api does that mean we no longer have support for SAML adapters on JavaEE? That's probably okay, but just wanted to double check on this one.
> 
> We're using `saml-core` and `saml-core-api` for adapters (JavaEE), and also for the server-side modules (JakartaEE). These SAML modules contain both JavaEE and JakartaEE dependencies and it should be possible to use them for both cases. AFAIK, there are no points in the code where it's necessary to require JavaEE or JakartaEE explicitly; especially for adapters. Thus, I presume we can leave it as is.
> 
> We covered that behavior for the WildFly adapter in the testsuite, which supports JavaEE, and everything works as expected.

Sounds good, thanks

**Review by stianst — APPROVED** [review]
Approving, as a note the follow-up tasks must be sorted out as a next priority once this is merged

**Review by miquelsi — APPROVED** [review]
Great job guys!

**Comment by mabartos** [comment]
Thank you all, guys! :partying_face:

**Pull request merged**
